### PR TITLE
Implement components representations in dag

### PR DESCRIPTION
**Pull request by Chiiizzzy**
#56 

Implemented a Graph representation as the following image:
GraphRepr

### CODE REVIEW

File: towhee/base_repr.py [hunk 1]
```diff
@@ -25,30 +22,32 @@ class BaseRepr:
...
-            value = self.__getattribute__(name)
-            if name[0] != '_' and not isinstance(value, types.FunctionType):
-                value = value.__class__.__name__
-                out += name + ': ' + str(value) + '\n'
-        return out
+    @name.setter
+    def name(self, value):
+        self._name = value
+
+    # def serialize(self) -> str:
```
Comment (fzliu) [inline towhee/base_repr.py:37]: This implementation of `serialize` is no longer correct based on our latest arch, remove this commented block and add a `raise NotImplementedError` instead.

File: towhee/dag/graph_repr.py [hunk 2]
```diff
@@ -21,16 +23,25 @@
...
-    are used during execution to load functions and pass data to the correct operators.
-    """
+    """This class presents a complete representation of a graph.
 
-    def __init__(self, name: str):
-        # TODO(Chiiizzzy)
+    A graph contains individual subcomponents, including Operators, Dataframes, and
+    Variables. Graph representations are used during execution to load functions and
+    pass data to the correct operators.
+    """
```
Comment (fzliu) [inline towhee/dag/graph_repr.py:31]: Add `__init__` argument descriptions in the class docstring here.

File: towhee/dag/graph_repr.py [hunk 3]
```diff
@@ -21,16 +23,25 @@
...
         self._operators = {}
         self._dataframes = {}
+        # if `file` is a loacl file
+        if file_or_url and os.path.isfile(file_or_url):
+            with open(file_or_url, 'r', encoding='utf-8') as f:
+                self._src = yaml.safe_load(f)
+            self.from_yaml(self._src)
+        # if `file` is a remote url
+        elif file_or_url:
+            self._src = requests.get(file_or_url).content.decode('utf-8')
```
Comment (fzliu) [inline towhee/dag/graph_repr.py:43]: Let's add a test case for remote file upload as well. As a source, we can create a gist for `towhee-io`.

File: towhee/dag/graph_repr.py [hunk 4]
```diff
@@ -40,15 +51,33 @@ def operators(self) -> Dict[str, OperatorRepr]:
...
+            raise ValueError('src is not a valid YAML file')
+        if not essentials.issubset(set(graph_dict.keys())):
+            raise ValueError('src cannot descirbe a DAG in Towhee')
+
+        # load name from YAML
+        self._name = graph_dict['graph']['name']
+
+        # load dataframes from YAML
+        for df in graph_dict['dataframes']:
+            self._dataframes[df['name']] = DataframeRepr(df['name'], str(df))
```
Comment (fzliu) [inline towhee/dag/graph_repr.py:76]: Are you sure we can use `str(df)` to transform this back into a YAML string? Isn't `df` a nested dictionary?

File: towhee/base_repr.py (hunk 1 above)
Comment (Chiiizzzy) [reply towhee/base_repr.py:37]: fixed

File: towhee/dag/graph_repr.py (hunk 2 above)
Comment (Chiiizzzy) [reply towhee/dag/graph_repr.py:31]: fixed

File: towhee/dag/graph_repr.py (hunk 4 above)
Comment (Chiiizzzy) [reply towhee/dag/graph_repr.py:76]: My bad, it loads dataframe from a string representation of a dict. Still, yaml.safe_load can load a dict from it

Comment (fzliu) [reply towhee/dag/graph_repr.py:76]: Weird, are you sure the resulting representation is still correct in all cases?

Comment (Chiiizzzy) [reply towhee/dag/graph_repr.py:76]: I've checked again and printed the result. It seems fine. In which case do you think will go wrong?

Comment (fzliu) [reply towhee/dag/graph_repr.py:76]: I am not 100% sure, but the output of `str(df)` is not YAML, there are probably cases where it will not work properly

Comment (Chiiizzzy) [reply towhee/dag/graph_repr.py]: fixed

File: towhee/base_repr.py [hunk 5]
```diff
@@ -25,30 +22,26 @@ class BaseRepr:
         name:
             Name of the internal object described by this representation.
     """
-
-    def __init__(self, name: str):
-        self._name = name
+    def __init__(self, name: str = None):
+        if name:
```
Comment (junjiejiangjjj) [inline towhee/base_repr.py:26]: if the name is None, just set self._name = None，make sure this attr is exist

File: towhee/base_repr.py [hunk 6]
```diff
@@ -25,30 +22,26 @@ class BaseRepr:
...
+            self._name = name
 
     @property
     def name(self):
         return self._name
 
+    @name.setter
+    def name(self, value):
```
Comment (junjiejiangjjj) [inline towhee/base_repr.py:34]: value: str

File: towhee/dag/graph_repr.py [hunk 7]
```diff
@@ -21,16 +23,31 @@
...
+    Variables. Graph representations are used during execution to load functions and
+    pass data to the correct operators.
+
+    Args:
+        name(`str`):
+            The representation name.
+        file_or_url(`str`):
+            The file or remote url that stores the information of this representation.
+    """
+    def __init__(self, name: str, file_or_url: str = None):
```
Comment (junjiejiangjjj) [inline towhee/dag/graph_repr.py:38]: We may support local file/HTTP/HDFs or other file systems, so we can use file_str, check types, 
HTTP:  file_str.low().startwiths("http")
HDFS: file_str.low().startwiths("hdfs")
local: file_str.low().startwiths('/') and so on. Write a function to check  the type and read the  file

File: towhee/dag/graph_repr.py [hunk 8]
```diff
@@ -40,15 +57,33 @@ def operators(self) -> Dict[str, OperatorRepr]:
     def dataframes(self) -> Dict[str, DataframeRepr]:
         return self._dataframes
 
-    def from_yaml(self, yaml: str):
+    def from_yaml(self, src: str):
```
Comment (junjiejiangjjj) [inline towhee/dag/graph_repr.py:60]: Use staticmethod, and return repr object

File: towhee/dag/operator_repr.py [hunk 9]
```diff
@@ -30,26 +29,53 @@ class OperatorRepr(BaseRepr):
...
+    def inputs(self, value: Dict[str, DataframeRepr]):
         self._df_in = value
 
-    @df_out.setter
-    def df_out(self, value: DataframeRepr):
+    @outputs.setter
+    def outputs(self, value: Dict[str, DataframeRepr]):
         self._df_out = value
+
+    def from_yaml(self, src: str, dataframes: Dict[str, DataframeRepr]):
```
Comment (junjiejiangjjj) [inline towhee/dag/operator_repr.py:60]: staticmethod, return a repr, can use another function to add  DatafremeRepr object

File: towhee/dag/variable_repr.py [hunk 10]
```diff
@@ -21,17 +20,16 @@ class VariableRepr(BaseRepr):
...
         dtype:
             A string or instance of `numpy.dtype` indicating the internal data type for
             this variable.
     """
-
-    def __init__(self, name: str, vtype: str, dtype: str):
-        super().__init__(name)
+    def __init__(self, vtype: str, dtype: str):
+        super().__init__()
         self._vtype = str(vtype)
```
Comment (junjiejiangjjj) [inline towhee/dag/variable_repr.py:33]: Input is already str type, why use str(...)

File: towhee/base_repr.py (hunk 5 above)
Comment (Chiiizzzy) [reply towhee/base_repr.py:26]: fixed

File: towhee/base_repr.py (hunk 6 above)
Comment (Chiiizzzy) [reply towhee/base_repr.py:34]: fixed

File: towhee/dag/graph_repr.py (hunk 7 above)
Comment (Chiiizzzy) [reply towhee/dag/graph_repr.py:38]: changed

File: towhee/dag/variable_repr.py (hunk 10 above)
Comment (Chiiizzzy) [reply towhee/dag/variable_repr.py:33]: fixed

File: towhee/dag/operator_repr.py (hunk 9 above)
Comment (Chiiizzzy) [reply towhee/dag/operator_repr.py]: fixed

File: towhee/dag/graph_repr.py (hunk 8 above)
Comment (Chiiizzzy) [reply towhee/dag/graph_repr.py]: fixed

File: towhee/dag/dataframe_repr.py [hunk 11]
```diff
@@ -16,79 +16,73 @@
...
-    def __setitem__(self, key: str, value: VariableRepr):
-        """Sets a single variable representation within the dataframe.
+        self._columns = []
+
+    @property
+    def columns(self):
+        return self._columns
+
+    @columns.setter
+    def columns(self, value: VariableRepr, index: int):
```
Comment (junjiejiangjjj) [inline towhee/dag/dataframe_repr.py:49]: Tow function named colums

File: towhee/dag/dataframe_repr.py [hunk 12]
```diff
@@ -16,79 +16,73 @@
...
-            was provided.
-        """
-        return self._columns.get(key)
-
-    def __setitem__(self, key: str, value: VariableRepr):
-        """Sets a single variable representation within the dataframe.
+        self._columns = []
+
+    @property
+    def columns(self):
```
Comment (junjiejiangjjj) [inline towhee/dag/dataframe_repr.py:45]: Implemented functions must have unit tests

**Review by junjiejiangjjj — CHANGES_REQUESTED** [review]
Implemented functions must have unit tests

**Comment by Chiiizzzy** [comment]
All problems fixed, added ut for each function @fzliu @junjiejiangjjj

File: towhee/dag/dataframe_repr.py [hunk 13]
```diff
@@ -12,83 +12,154 @@
...
+        return self._columns
+
+    @columns.setter
+    def columns(self, value: List[VariableRepr]):
+        self._columns = value
 
-    def __getitem__(self, key) -> VariableRepr:
-        """Access a variable representation via dictionary-like indexing.
+    @staticmethod
+    def is_format(info: list):
```
Comment (junjiejiangjjj) [inline towhee/dag/dataframe_repr.py:50]: return True/False

**Review by junjiejiangjjj — APPROVED** [review]
/approve
/lgtm

**Pull request merged**